sql/log.cc

```diff
@@ -6691,6 +6702,85 @@ bool THD::binlog_write_table_maps()
+
+    - Second pass: Writes the actual ANNOTATE and TABLE_MAP events
+
+  */
+  if (binlog_dump_filter->is_on() || !binlog_dump_filter->is_db_empty()) 
+  {
+      for (MYSQL_LOCK **cur_lock= locks ; cur_lock < locks_end ; cur_lock++)
+      {
+        TABLE **const end_ptr= (*cur_lock)->table + (*cur_lock)->table_count;
+        for (TABLE **table_ptr= (*cur_lock)->table;
+            table_ptr != end_ptr ;
+            ++table_ptr)
+        {
+          TABLE *table= *table_ptr;
+
+          const char* db= table->s->db.str;
+          bool should_replicate_db= binlog_dump_filter->db_ok(db);
+          bool should_replicate_table= 
+            binlog_dump_filter->table_ok(db, table->alias.c_ptr());
+
+          /*
+            If there is no rule set the default returned would be true
+            so if there are no rules for db it would return true regardless of
+            the table rules so by having false in the one that actually contains
+            the rules then this means that this table should NOT be filtered 
+            so the only possible choice to avoid the default being true is 
+            to make sure both flags are true using an && instead of an ||
+
+            Otherwise then this should be filtered partially by using
+            the cache's filtered table ids .. This step is useful in order
+            to be able to filter the *_ROWS_EVENT corresponding to TABLE_MAP
+            later in case there is a multi-statement qury that affects
+            multiple tables 
+            e.g: TABLE_MAP1, TABLE_MAP2 are written and then two 
+            WRITE_ROWS_EVENT , if the first one is the only one
+             that has to be filtered then there has to be a way
+             to know from the table_id of WRITE_ROWS_EVENT 
+          */
+          bool should_replicate= should_replicate_db && should_replicate_table;
+          if(should_replicate)
+          {
+            found_unfiltered_table= true;
+          }
+          else {
+            found_filtered_table= true;
+            /*
+              TODO: should this be done only if filter_all_tables is false ? 
+            */
+            bool is_transactional= table->file->row_logging_has_trans;
+            is_transactional|= this->variables.option_bits & OPTION_GTID_BEGIN;
+
+            binlog_cache_mngr *const cache_mngr= this->binlog_get_cache_mngr();
+            binlog_cache_data *cache_data= (cache_mngr->
+                                  get_binlog_cache_data(is_transactional));
+            ulonglong table_id= table->s->table_map_id;
+            uchar *stored= (uchar*) my_malloc(PSI_INSTRUMENT_ME,
+                                    sizeof(ulonglong), MYF(0));
+            memcpy(stored, &table_id, sizeof(ulonglong));
+
+            my_hash_insert(&cache_data->partial_filtered_table_ids, stored);
+            
+          }
+        }
+      }
+    filter_all_tables= found_filtered_table && !found_unfiltered_table;
```

Comment:
Do we need this `filter_all_tables` variable (and parameter to `write_table_map`? Can we just set `cache_data->event_group_rpl_filter= true` here? Then use that going forward.